ppa/archive/management/commands/index.py

```diff
@@ -0,0 +1,154 @@
+
+        if self.options['source_ids']:
+            works = works.filter(source_id__in=self.options['source_ids'])
+
+        total_to_index = 0
+        # calculate total to index based on what we are indexing
+        if self.options['index'] in ['works', 'all']:
+            # total works to be indexed
+            total_to_index += works.count()
+        if self.options['index'] in ['pages', 'all']:
+            # total pages to be indexed
+            # NOTE: should always be a value if id is valid, but fallback to zero for None
+            total_to_index += works.aggregate(total_pages=Sum('page_count'))['total_pages'] or 0
+
+        # initialize progressbar if requested and indexing more than 5 items
+        progbar = None
+        if not self.options['no_progress'] and total_to_index > 5:
+            progbar = progressbar.ProgressBar(redirect_stdout=True,
+                                              max_value=total_to_index)
+
+        # index works
+        count = 0
+        if self.options['index'] in ['works', 'all']:
+            # use django paginator to index chunks of works at a time
+            paginator = Paginator(works, 100)
+            for page in range(1, paginator.num_pages + 1):
+                self.index([work.index_data() for work in paginator.page(page).object_list])
+
+                count += paginator.page(page).object_list.count()
+                if progbar:
+                    progbar.update(count)
+
+        self.solr.commit(self.solr_collection)
+
+        # index pages for each work
+        if self.options['index'] in ['pages', 'all']:
+            for work in works:
+                # page index data returns a generator
+                page_data = work.page_index_data()
+
+                # iterate over the generator and index in chunks
+                page_chunk = list(itertools.islice(page_data, 150))
+                while page_chunk:
+                    self.index(page_chunk)
+                    page_chunk = list(itertools.islice(page_data, 150))
+
+                # for simplicity, update progress bar for each work
+                # rather than each chunk of pages
+                count += work.page_count
+                if progbar:
+                    progbar.update(count)
+
+        self.solr.commit(self.solr_collection)
+
+    def index(self, index_data):
+        '''index an iterable into the configured solr instance
+        and solr collection'''
+
+        # NOTE: currently no good way to catch a connection
+        # error when Solr is not running because we get multiple
+        # connections during handling of exceptions.
+        try:
+            self.solr.index(self.solr_collection, index_data,
+                            params=self.solr_index_opts)
+        except Exception as err:
+        # except (ConnectionError, RequestException) as err:
+            # NOTE: this is still pretty ugly; what part should we return?
+            raise CommandError(err)
```

Comment:
does it have the gnarly java stacktrace? that seems like something we probably don't need (all of). Conversely, the underlying java exception class (e.g. is it an OOMError or not) is probably good to see

Reply:
I would tend to argue for a full stacktrace, gnarly though it be (Java/Scala, here's looking at you).

Reply:
The problem is that there are multiple exceptions happening in the processing of other connections - this is also why I had such a hard time figuring out the right exception to even catch.  There seem to be some in SolrClient, some in requests, and some in urllib3.